language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs

```diff
@@ -104,222 +135,266 @@ kindOfTypeBuiltin TyByteString = sizeToType
+kindCheckM x ty k = do
+    tyK <- kindOf ty
+    when (tyK /= k) $ throwError (KindMismatch x (void ty) k tyK)
+
 -- | Check that the kind of a pattern functor is @(k -> *) -> k -> *@.
 kindCheckPatternFunctorM
     :: ann
-    -> Type TyNameWithKind ann  -- ^ A pattern functor.
-    -> Kind ()                  -- ^ @k@.
+    -> Type TyName ann  -- ^ A pattern functor.
+    -> Kind ()          -- ^ @k@.
     -> TypeCheckM ann ()
 kindCheckPatternFunctorM ann pat k =
     kindCheckM ann pat $ KindArrow () (KindArrow () k (Type ())) (KindArrow () k (Type ()))
 
--- | Check a 'Type' against a 'Kind'.
-kindCheckM :: a -> Type TyNameWithKind a -> Kind () -> TypeCheckM a ()
-kindCheckM x ty k = do
-    tyK <- kindOf ty
-    when (tyK /= k) $ throwError (KindMismatch x (void ty) k tyK)
-
 -- | Apply a 'TypeBuiltin' to a 'Size' and wrap in 'NormalizedType'.
 applySizedNormalized :: TypeBuiltin -> Size -> NormalizedType tyname ()
 applySizedNormalized tb = NormalizedType . TyApp () (TyBuiltin () tb) . TyInt ()
 
 dummyUnique :: Unique
 dummyUnique = Unique 0
 
-dummyTyName :: TyNameWithKind ()
-dummyTyName = TyNameWithKind (TyName (Name ((), Type ()) "*" dummyUnique))
+dummyTyName :: TyName ()
+dummyTyName = TyName (Name () "*" dummyUnique)
 
 dummyKind :: Kind ()
 dummyKind = Type ()
 
-dummyType :: Type TyNameWithKind ()
+dummyType :: Type TyName ()
 dummyType = TyVar () dummyTyName
 
 -- | Look up a 'DynamicBuiltinName' in the 'DynBuiltinNameTypes' environment.
-lookupDynamicBuiltinName :: a -> DynamicBuiltinName -> TypeCheckM a (NormalizedType TyNameWithKind ())
+lookupDynamicBuiltinName :: a -> DynamicBuiltinName -> TypeCheckM a (NormalizedType TyName ())
 lookupDynamicBuiltinName ann name = do
     dbnts <- asks $ unDynamicBuiltinNameTypes . _typeConfigDynBuiltinNameTypes
     case Map.lookup name dbnts of
-        Nothing    ->
+        Nothing ->
             throwError $ UnknownDynamicBuiltinName ann (UnknownDynamicBuiltinNameErrorE name)
-        Just quoTy -> do
-            ty <- liftQuote quoTy
-            annotateNormalizeType ann (DynBuiltinName () name) ty
+        Just ty -> rename ty
 
 -- | Get the 'Type' of a 'Constant' wrapped in 'NormalizedType'.
-typeOfConstant :: Constant a -> NormalizedType TyNameWithKind ()
+typeOfConstant :: Constant a -> NormalizedType TyName ()
 typeOfConstant (BuiltinInt  _ size _) = applySizedNormalized TyInteger    size
 typeOfConstant (BuiltinBS   _ size _) = applySizedNormalized TyByteString size
 typeOfConstant (BuiltinSize _ size)   = applySizedNormalized TySize       size
 typeOfConstant (BuiltinStr _ _)       = NormalizedType $ TyBuiltin () TyString
 
-typeOfBuiltin :: Builtin a -> TypeCheckM a (NormalizedType TyNameWithKind ())
-typeOfBuiltin (BuiltinName    ann name) = normalizedAnnotatedTypeOfBuiltinName ann name
+typeOfBuiltin :: Builtin a -> TypeCheckM a (NormalizedType TyName ())
+-- We have a weird corner case here: the type of a 'BuiltinName' can contain 'TypedBuiltinDyn', i.e.
```

Comment:
This can't actually happen, right? It's just that we can't statically show that it won't?

Reply:
This can happen if someone adds such a static builtin. Or turns `Bool` into a dynamic builtin type (which is what I want to do at some point).

Reply:
Right, I meant currently. We could do a similar thing and enforce that the types of static builtins must be normalized, but that's a similar annoyance as below (although I don't think it would be that bad).

(Why do you want to make `Bool` into a dynamic builtin? I don't think it buys us much.)

Reply:
> Right, I meant currently. 

Yeah, currently can't happen.

> We could do a similar thing and enforce that the types of static builtins must be normalized, but that's a similar annoyance as below (although I don't think it would be that bad).

As the comment says, the types of static builtin names can mention dynamic builtin types. That's what is weird about this corner case, so it's exactly the same annoyance.

The dynamic builtins story is a little bit of a mess, I admit. It grew chaotically for two distinct use-cases and under time pressure, it needs to be improved somehow.

Reply:
Note that in PLC produced from the Plutus Tx compiler there are comparatively few PLC booleans - we more or less eliminate them immediately in favour of the booleans compiled from Haskell. So I don't think it's worth spending that much effort making them work better.